File: packages/ts/tests/mock_golden.test.ts
// E3.9 — golden vectors for the offline mock, shared with Python and Rust.
// The fixture is the same file the other runtimes read: any drift in one runtime breaks parity
// here instead of shipping a false claim.
import { test } from "node:test";
import * as assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import * as path from "node:path";
import {
  JevClient,
  MOCK_CHOICE_BEST_CONFLICT,
  MOCK_CHOICE_BEST_PEAKED,
  MOCK_SCORE_BEST_PEAKED,
  mockDistribution,
} from "../src/client.js";

const fixture = JSON.parse(
  readFileSync(path.join(process.cwd(), "..", "..", "tests", "fixtures", "mock_golden.json"), "utf-8")
);
const client = new JevClient({ forceMock: true });

function choiceFor(state: string) {
  const answer = (client as any).simulateSystemOne(
    state,
    { category: { type: "choice", instructions: "What is the root failure type in this error trace?", criteria: fixture.choice_criteria } },
    client.model
  ).answers["category"];
  return answer;
}

function nearly(actual: number, expected: number, label: string) {
  assert.ok(Math.abs(actual - expected) < 1e-9, `${label}: ${actual} != ${expected}`);
}

test("E3.9 mock constants match the cross-runtime contract", () => {
  assert.equal(MOCK_CHOICE_BEST_PEAKED, 0.85);
  assert.equal(MOCK_CHOICE_BEST_CONFLICT, 0.55);
  assert.equal(MOCK_SCORE_BEST_PEAKED, 0.8);
});

test("E3.9 peaked choice distribution matches the golden vector", () => {
  const answer = choiceFor(fixture.peaked.state);
  assert.equal(answer.choice, fixture.peaked.expected.choice);
  for (const [key, value] of Object.entries<number>(fixture.peaked.expected.probabilities)) {
    nearly(answer.probabilities[key], value, `peaked ${key}`);
  }
  assert.equal(Object.keys(answer.probabilities).length, Object.keys(fixture.choice_criteria).length);
});

test("E3.9 conflicting signals lower the peak (deterministic escalation)", () => {
  const answer = choiceFor(fixture.conflict.state);
  assert.equal(answer.choice, fixture.conflict.expected.choice);
  for (const [key, value] of Object.entries<number>(fixture.conflict.expected.probabilities)) {
    nearly(answer.probabilities[key], value, `conflict ${key}`);
  }
  nearly(answer.probabilities[answer.choice], MOCK_CHOICE_BEST_CONFLICT, "conflict peak");
});

test("E3.9 distributions sum to one in every vector", () => {
  for (const name of ["peaked", "conflict"] as const) {
    const answer = choiceFor(fixture[name].state);
    const total = Object.values<number>(answer.probabilities).reduce((a, b) => a + b, 0);
    nearly(total, 1.0, `${name} sum`);
  }
});

test("E3.9 score distribution is exposed and matches the golden vector", () => {
  const answer = (client as any).simulateSystemOne(
    fixture.score.state,
    { severity: { type: "score", instructions: "How severe is this failure?", criteria: fixture.score.levels } },
    client.model
  ).answers["severity"];
  nearly(answer.score, fixture.score.expected.score, "score");
  for (const [key, value] of Object.entries<number>(fixture.score.expected.probabilities)) {
    nearly(answer.probabilities[key], value, `score ${key}`);
  }
  nearly(
    Object.values<number>(answer.probabilities).reduce((a, b) => a + b, 0),
    1.0,
    "score sum"
  );
});

test("E3.9 noul stays a scalar and mockDistribution handles degenerate inputs", () => {
  const answer = (client as any).simulateSystemOne(
    fixture.noul.state,
    { skip_llm: { type: "noul", instructions: fixture.noul.instruction } },
    client.model
  ).answers["skip_llm"];
  nearly(answer.noul, fixture.noul.expected.noul, "noul");
  assert.equal(answer.probabilities, undefined, "Noul has no distribution by contract");

  assert.deepEqual(mockDistribution(["only"], "only", 0.85), { only: 1.0 });
  const two = mockDistribution(["a", "b"], "b", 0.5);
  nearly(two.a, 0.5, "two a");
  nearly(two.b, 0.5, "two b");
});

File: packages/rust/tests/mock_golden_test.rs
//! E3.9 — golden vectors for the offline mock, shared with Python and TypeScript.
//! The fixture is the same file the other runtimes read: any drift in one runtime breaks parity
//! here instead of shipping a false claim.
use jev_harness::client::{
    MOCK_CHOICE_BEST_CONFLICT, MOCK_CHOICE_BEST_PEAKED, MOCK_SCORE_BEST_PEAKED,
};
use jev_harness::types::{ChoiceQuestion, NoulQuestion, Question, ScoreQuestion};
use jev_harness::JevClient;
use serde_json::Value;
use std::collections::HashMap;

const FIXTURE: &str = include_str!("../../../tests/fixtures/mock_golden.json");

fn fixture() -> Value {
    serde_json::from_str(FIXTURE).expect("golden fixture is valid JSON")
}

fn criteria(value: &Value) -> HashMap<String, String> {
    value
        .as_object()
        .expect("criteria object")
        .iter()
        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or_default().to_string()))
        .collect()
}

fn choice_answer(state: &str) -> (String, HashMap<String, f64>) {
    let fx = fixture();
    let mut questions = HashMap::new();
    questions.insert(
        "category".to_string(),
        Question::Choice(ChoiceQuestion {
            instructions: "What is the root failure type in this error trace?".into(),
            criteria: criteria(&fx["choice_criteria"]),
        }),
    );
    let client = JevClient::with_mock();
    let resp = client.simulate_system_one(state, &questions, &client.model);
    match resp.answers.get("category").expect("category answer") {
        jev_harness::types::Answer::Choice(a) => (
            a.choice.clone(),
            a.probabilities.clone().unwrap_or_default(),
        ),
        other => panic!("expected a choice answer, got {:?}", other),
    }
}

fn nearly(actual: f64, expected: f64, label: &str) {
    assert!(
        (actual - expected).abs() < 1e-9,
        "{label}: {actual} != {expected}"
    );
}

#[test]
fn mock_constants_match_the_cross_runtime_contract() {
    nearly(MOCK_CHOICE_BEST_PEAKED, 0.85, "peaked");
    nearly(MOCK_CHOICE_BEST_CONFLICT, 0.55, "conflict");
    nearly(MOCK_SCORE_BEST_PEAKED, 0.80, "score");
}

#[test]
fn peaked_choice_distribution_matches_the_golden_vector() {
    let fx = fixture();
    let (choice, probs) = choice_answer(fx["peaked"]["state"].as_str().unwrap());
    assert_eq!(choice, fx["peaked"]["expected"]["choice"].as_str().unwrap());
    assert_eq!(
        probs.len(),
        fx["choice_criteria"].as_object().unwrap().len(),
        "every option must carry a probability"
    );
    for (key, value) in fx["peaked"]["expected"]["probabilities"]
        .as_object()
        .unwrap()
    {
        nearly(probs[key], value.as_f64().unwrap(), key);
    }
}

#[test]
fn conflicting_signals_lower_the_peak() {
    let fx = fixture();
    let (choice, probs) = choice_answer(fx["conflict"]["state"].as_str().unwrap());
    assert_eq!(
        choice,
        fx["conflict"]["expected"]["choice"].as_str().unwrap()
    );
    for (key, value) in fx["conflict"]["expected"]["probabilities"]
        .as_object()
        .unwrap()
    {
        nearly(probs[key], value.as_f64().unwrap(), key);
    }
    nearly(probs[&choice], MOCK_CHOICE_BEST_CONFLICT, "conflict peak");
    assert!(probs[&choice] < MOCK_CHOICE_BEST_PEAKED);
}

#[test]
fn distributions_sum_to_one() {
    let fx = fixture();
    for name in ["peaked", "conflict"] {
        let (_choice, probs) = choice_answer(fx[name]["state"].as_str().unwrap());
        let total: f64 = probs.values().sum();
        nearly(total, 1.0, name);
    }
}

#[test]
fn score_distribution_is_exposed() {
    let fx = fixture();
    let levels: Vec<String> = fx["score"]["levels"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap().to_string())
        .collect();
    let mut questions = HashMap::new();
    questions.insert(
        "severity".to_string(),
        Question::Score(ScoreQuestion {
            instructions: "How severe is this failure?".into(),
            criteria: levels.clone(),
        }),
    );
    let client = JevClient::with_mock();
    let resp = client.simulate_system_one(
        fx["score"]["state"].as_str().unwrap(),
        &questions,
        &client.model,
    );
    let answer = match resp.answers.get("severity").expect("severity answer") {
        jev_harness::types::Answer::Score(a) => a,
        other => panic!("expected a score answer, got {:?}", other),
    };
    nearly(
        answer.score,
        fx["score"]["expected"]["score"].as_f64().unwrap(),
        "score",
    );
    let probs = answer
        .probabilities
        .as_ref()
        .expect("score probabilities exposed");
    assert_eq!(probs.len(), levels.len(), "one probability per level");
    for (key, value) in fx["score"]["expected"]["probabilities"]
        .as_object()
        .unwrap()
    {
        nearly(probs[key], value.as_f64().unwrap(), key);
    }
    nearly(probs.values().sum::<f64>(), 1.0, "score sum");
}

#[test]
fn noul_stays_a_scalar() {
    let fx = fixture();
    let mut questions = HashMap::new();
    questions.insert(
        "skip_llm".to_string(),
        Question::Noul(NoulQuestion {
            instructions: fx["noul"]["instruction"].as_str().unwrap().to_string(),
        }),
    );
    let client = JevClient::with_mock();
    let resp = client.simulate_system_one(
        fx["noul"]["state"].as_str().unwrap(),
        &questions,
        &client.model,
    );
    match resp.answers.get("skip_llm").expect("noul answer") {
        jev_harness::types::Answer::Noul(a) => nearly(
            a.noul,
            fx["noul"]["expected"]["noul"].as_f64().unwrap(),
            "noul",
        ),
        other => panic!("expected a noul answer, got {:?}", other),
    }
}
